nemo/collections/asr/data/audio_to_text.py

```diff
@@ -208,6 +226,125 @@ def expand_audio_filepaths(audio_tar_filepaths, shard_strategy: str, world_size:
+                    logging.info('Cached at: %s', str(cached_manifest_file))
+
+                    if cache_audio:
+                        logging.info('Cache audio from manifest file: %s', manifest_file)
+                        # Assumes that manifest is using relative paths
+                        manifest_dir = os.path.dirname(manifest_file)
+                        # Prepare all store objects
+                        audio_objects = []
+                        with open(cached_manifest_file, 'r') as f:
+                            for line in f:
+                                item = json.loads(line)
+                                store_path = os.path.join(manifest_dir, item['audio_filepath'])
+                                audio_objects.append(StoreObject(store_path=store_path))
+
+                        if num_workers is not None and num_workers > 1:
+                            logging.debug('Using multiprocessing with num_workers: %d.', num_workers)
+                            with multiprocessing.Pool(processes=num_workers) as p:
+                                result = list(tqdm(p.imap(store_object_get, audio_objects), total=len(audio_objects)))
+                        else:
+                            logging.debug('Using a single process.')
+                            result = []
+                            for audio_object in tqdm(audio_objects):
+                                result.append(audio_object.get() is not None)
+
+                        if any(result) == False:
+                            raise RuntimeError('Some files not downloaded successfully')
+                        logging.info('Caching complete')
+
+                else:
+                    # Nothing to do here
+                    logging.debug('Manifest is not on a store: %s', manifest_file)
+
+        # Handle distributed environment
+        if shared_cache is None:
+            shared_cache = is_store_cache_shared()
+
+        if shared_cache:
+            logging.debug('Cache is shared among nodes.')
+            is_rank_zero = is_global_rank_zero()
+        else:
+            logging.debug('Cache is not shared among nodes.')
+            local_rank = int(os.environ.get("LOCAL_RANK", 0))
+            is_rank_zero = local_rank == 0
+
+        if torch.distributed.is_available() and torch.distributed.is_initialized():
+            logging.debug('Distributed environment is available and initialized.')
+            if is_rank_zero:
+                logging.info('Cache data from %s rank 0', 'global' if shared_cache else 'local')
+                cache_data(
+                    manifest_filepaths=manifest_filepaths,
+                    cache_audio=cache_audio,
+                    num_workers=num_workers,
+                    max_num_workers=max_num_workers,
+                )
+            logging.debug('Reached barrier')
+            torch.distributed.barrier()
+
+        elif is_rank_zero:
```

Comment:
Can we fully remove this part and raise an error in torch distributed is not initialized? This seems quite fragile to me with the ad-hoc sleep time, and potentially undefined behavior if multiple processes write to the same file. So I'd prefer to require distributed to be initialized for ais to be used.

And for the rank, would it be safer to use `torch.distributed.get_rank` (if we assume it's always initialized)? It seems that `is_global_rank_zero` checks a bunch of environment variables which (I guess) might not always work correctly